File: P250136.py
from collections import deque

dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

def solution(land):

    n, m = len(land), len(land[0])
    answer = [0] * m
    visited = [[0 for _ in range(m)] for _ in range(n)]
    for j in range(m):
        for i in range(n):
            if land[i][j] and not visited[i][j]:
                queue = deque()
                queue.append([i, j])
                visited[i][j] = 1
                cnt = 0
                l, r = j, j
                
                while queue:
                    node = queue.popleft()
                    cnt += 1
                    a, b = node[0], node[1]
                    if b < l:
                        l = b
                    if b > r:
                        r = b
                    
                    for k in range(4):
                        x, y = a + dx[k], b + dy[k]
                        
                        if 0 <= x < n and 0 <= y < m and not visited[x][y] and land[x][y]:
                            visited[x][y] = 1
                            queue.append([x, y])
                
                for k in range(l, r+1):
                    answer[k] += cnt
    
    return max(answer)

File: P250136.java
import java.util.ArrayDeque;

public class P250136 {
    int[] dx = new int[] {1, -1, 0, 0};
    int[] dy = new int[] {0, 0, 1, -1};
    
    public int solution(int[][] land) {
        
        int n = land.length;
        int m = land[0].length;
        int[] answer = new int[m];
        int[][] visited = new int[n][m];
        for (int j=0; j<m; j++) {
            for (int i=0; i<n; i++) {
                if (land[i][j] != 0 && visited[i][j] == 0) {
                    ArrayDeque<int[]> queue = new ArrayDeque<>();
                    queue.add(new int[] {i, j});
                    visited[i][j] = 1;
                    int cnt = 0;
                    int l = j;
                    int r = j;
                    
                    while (!queue.isEmpty()) {
                        int[] node = queue.pollFirst();
                        cnt += 1;
                        int a = node[0];
                        int b = node[1];
                        if (b < l) l = b;
                        if (b > r) r = b;
                        
                        for (int k=0; k<4; k++) {
                            int x = a + this.dx[k];
                            int y = b + this.dy[k];
                            
                            if (0 <= x && x < n && 0 <= y && y < m && visited[x][y] == 0 && land[x][y] != 0) {
                                visited[x][y] = 1;
                                queue.add(new int[] {x, y});
                            }
                        }
                    }
                    for (int k=l; k<r+1; k++) answer[k] += cnt;
                }
            }
        }
        int Max = 0;
        for (int i=0; i<answer.length; i++) if (answer[i] > Max) Max = answer[i];
        return Max;
    }
}
